
# Weather Forecast

Here considering given data set, compiled over the 10000 epochs finally able to get description of the data set.

loss: -7.5724 

accuracy: 0.0993






In [2]:
import numpy as np
import pandas as pd
import tensorflow as tf
from tensorflow import keras
from matplotlib import pyplot as plt
from sklearn import preprocessing
%matplotlib inline


In [3]:
!git clone 'https://github.com/sachinkavindaa/Machine-learning-with-Python'

Cloning into 'Machine-learning-with-Python'...
remote: Enumerating objects: 77, done.
remote: Counting objects: 100% (77/77), done.
remote: Compressing objects: 100% (46/46), done.
remote: Total 77 (delta 18), reused 71 (delta 12), pack-reused 0
Unpacking objects: 100% (77/77), done.


In [4]:
df = pd.read_csv('/content/Machine-learning-with-Python/ML/Weather/Weather_3/Weather.csv')
df.head()

Date  T_Max  T_Min  Wind  ...  Humidity  Precip  Pressure        Weather
0  1/1/2010     30     22     7  ...        69     0.0      1011          Sunny
1  1/2/2009     30     22     7  ...        68     0.0      1012          Sunny
2  1/3/2009     30     22     7  ...        68     0.0      1012          Sunny
3  1/4/2009     31     22     7  ...        73     0.0      1011          Sunny
4  1/5/2009     30     23     7  ...        76     0.4      1010  Partly couldy

[5 rows x 9 columns]

In [5]:
merged = df[['T_Max', 'T_Min', 'Wind', 'Could','Precip','Humidity','Pressure']]
merged.head()

T_Max  T_Min  Wind  Could  Precip  Humidity  Pressure
0     30     22     7      4     0.0        69      1011
1     30     22     7      3     0.0        68      1012
2     30     22     7      1     0.0        68      1012
3     31     22     7      8     0.0        73      1011
4     30     23     7     29     0.4        76      1010

In [6]:

def min_max_scaling(merged):
    # copy the dataframe
    df_norm = merged.copy()
    # apply min-max scaling
    for column in df_norm.columns:
        df_norm[column] = (df_norm[column] - df_norm[column].min()) / (df_norm[column].max() - df_norm[column].min())
        
    return df_norm
    
# call the min_max_scaling function
df_scaled = min_max_scaling(merged)

df_scaled


T_Max     T_Min      Wind     Could    Precip  Humidity  Pressure
0     0.88  0.004425  0.190476  0.031915  0.000000  0.178571  0.996714
1     0.88  0.004425  0.190476  0.021277  0.000000  0.142857  0.997809
2     0.88  0.004425  0.190476  0.000000  0.000000  0.142857  0.997809
3     0.92  0.004425  0.190476  0.074468  0.000000  0.321429  0.996714
4     0.88  0.008850  0.190476  0.297872  0.035088  0.428571  0.995619
..     ...       ...       ...       ...       ...       ...       ...
360   0.84  0.008850  0.095238  0.500000  0.000000  0.785714  0.996714
361   0.84  0.013274  0.095238  0.574468  0.000000  0.785714  0.996714
362   0.84  1.000000  0.190476  0.531915  0.008772  0.750000  0.996714
363   0.84  0.008850  0.047619  0.712766  0.219298  0.821429  0.996714
364   0.88  0.008850  0.047619  0.170213  0.061404  0.607143  0.995619

[365 rows x 7 columns]

In [7]:
df_scaled.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 365 entries, 0 to 364
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   T_Max     365 non-null    float64
 1   T_Min     365 non-null    float64
 2   Wind      365 non-null    float64
 3   Could     365 non-null    float64
 4   Precip    365 non-null    float64
 5   Humidity  365 non-null    float64
 6   Pressure  365 non-null    float64
dtypes: float64(7)
memory usage: 20.1 KB


In [8]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(df_scaled[['T_Max', 'T_Min', 'Wind', 'Could','Humidity','Pressure']],df_scaled.Precip,test_size =0.2,random_state=25)

In [9]:
X_train

T_Max     T_Min      Wind     Could  Humidity  Pressure
164   0.84  0.017699  0.619048  0.234043  0.678571  0.997809
125   0.84  0.022124  0.380952  0.500000  0.750000  0.995619
73    0.92  0.013274  0.190476  0.159574  0.571429  0.996714
133   0.88  0.022124  0.666667  0.436170  0.642857  0.992333
31    0.92  0.008850  0.190476  0.223404  0.285714  0.997809
..     ...       ...       ...       ...       ...       ...
255   0.84  0.017699  0.523810  0.414894  0.678571  0.997809
317   0.84  0.013274  0.380952  0.617021  0.750000  0.993428
143   0.76  0.017699  1.000000  0.361702  0.892857  0.991238
318   0.72  0.013274  0.190476  0.755319  0.964286  0.994524
132   0.88  0.022124  0.619048  0.372340  0.607143  0.992333

[292 rows x 6 columns]

In [10]:
df_scaled.shape


(365, 7)

In [11]:
len(X_train)

292

In [12]:
X_train_scaled = X_train.copy()
X_train_scaled[['T_Max', 'T_Min', 'Wind', 'Could','Humidity','Pressure']]= X_train_scaled[['T_Max', 'T_Min', 'Wind', 'Could','Humidity','Pressure']]

X_test_scaled = X_test.copy()
X_test_scaled[['T_Max', 'T_Min', 'Wind', 'Could','Humidity','Pressure']]= X_test_scaled[['T_Max', 'T_Min', 'Wind', 'Could','Humidity','Pressure']]



In [13]:
X_train_scaled



T_Max     T_Min      Wind     Could  Humidity  Pressure
164   0.84  0.017699  0.619048  0.234043  0.678571  0.997809
125   0.84  0.022124  0.380952  0.500000  0.750000  0.995619
73    0.92  0.013274  0.190476  0.159574  0.571429  0.996714
133   0.88  0.022124  0.666667  0.436170  0.642857  0.992333
31    0.92  0.008850  0.190476  0.223404  0.285714  0.997809
..     ...       ...       ...       ...       ...       ...
255   0.84  0.017699  0.523810  0.414894  0.678571  0.997809
317   0.84  0.013274  0.380952  0.617021  0.750000  0.993428
143   0.76  0.017699  1.000000  0.361702  0.892857  0.991238
318   0.72  0.013274  0.190476  0.755319  0.964286  0.994524
132   0.88  0.022124  0.619048  0.372340  0.607143  0.992333

[292 rows x 6 columns]

In [14]:
model = keras.Sequential([
                          keras.layers.Dense(1,input_shape=(6,),activation='sigmoid',kernel_initializer='ones',bias_initializer='zeros')
])

In [15]:
model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

model.fit(X_train_scaled,y_train,epochs=300)

Epoch 1/300
10/10 [==============================] - 0s 3ms/step - loss: 3.1032 - accuracy: 0.0034
Epoch 2/300
10/10 [==============================] - 0s 3ms/step - loss: 3.0650 - accuracy: 0.0034
Epoch 3/300
10/10 [==============================] - 0s 3ms/step - loss: 3.0266 - accuracy: 0.0034
Epoch 4/300
10/10 [==============================] - 0s 3ms/step - loss: 2.9883 - accuracy: 0.0034
Epoch 5/300
10/10 [==============================] - 0s 3ms/step - loss: 2.9503 - accuracy: 0.0034
Epoch 6/300
10/10 [==============================] - 0s 3ms/step - loss: 2.9123 - accuracy: 0.0034
Epoch 7/300
10/10 [==============================] - 0s 3ms/step - loss: 2.8745 - accuracy: 0.0034
Epoch 8/300
10/10 [==============================] - 0s 3ms/step - loss: 2.8368 - accuracy: 0.0034
Epoch 9/300
10/10 [==============================] - 0s 3ms/step - loss: 2.7994 - accuracy: 0.0034
Epoch 10/300
10/10 [==============================] - 0s 3ms/step - loss: 2.7619 - accuracy: 0.0034
Epoch 11/

In [16]:
X_test_scaled 

T_Max     T_Min      Wind     Could  Humidity  Pressure
364   0.88  0.008850  0.047619  0.170213  0.607143  0.995619
225   0.84  0.022124  0.761905  0.563830  0.714286  0.994524
33    0.88  0.008850  0.190476  0.053191  0.071429  0.997809
100   0.88  0.013274  0.142857  0.372340  0.750000  0.992333
45    0.92  0.004425  0.142857  0.244681  0.321429  0.995619
..     ...       ...       ...       ...       ...       ...
149   0.80  0.017699  0.761905  0.808511  0.857143  0.994524
192   0.88  0.022124  0.619048  0.106383  0.500000  0.995619
287   0.88  0.013274  0.190476  0.404255  0.607143  0.996714
168   0.76  0.017699  0.809524  0.744681  0.785714  0.994524
341   0.84  0.013274  0.047619  0.585106  0.857143  0.995619

[73 rows x 6 columns]

In [17]:
model.evaluate(X_test_scaled,y_test)

3/3 [==============================] - 0s 3ms/step - loss: 0.2495 - accuracy: 0.3425


[0.2494955211877823, 0.34246575832366943]

In [18]:
model.predict(X_test_scaled)

array([[0.10696125],
       [0.07982428],
       [0.11637107],
       [0.09764025],
       [0.10763293],
       [0.08444721],
       [0.07060355],
       [0.11095428],
       [0.10259119],
       [0.07619622],
       [0.08404898],
       [0.08998002],
       [0.08577264],
       [0.08477487],
       [0.08824079],
       [0.07208892],
       [0.07736523],
       [0.09797491],
       [0.10468846],
       [0.09259775],
       [0.09647424],
       [0.08788701],
       [0.09347907],
       [0.08659996],
       [0.09614702],
       [0.10329697],
       [0.08896136],
       [0.12137064],
       [0.08022577],
       [0.09571616],
       [0.08236077],
       [0.08278839],
       [0.09412871],
       [0.09416921],
       [0.09021119],
       [0.0988396 ],
       [0.08005469],
       [0.09704101],
       [0.09136397],
       [0.09675517],
       [0.11020066],
       [0.08737711],
       [0.0869054 ],
       [0.09808882],
       [0.1016616 ],
       [0.0771935 ],
       [0.08647314],
       [0.080

In [19]:
y_test

364    0.061404
225    0.052632
33     0.000000
100    0.000000
45     0.017544
         ...   
149    0.043860
192    0.000000
287    0.175439
168    0.219298
341    0.315789
Name: Precip, Length: 73, dtype: float64

In [20]:
coef, intercept = model.get_weights()
coef, intercept

(array([[-0.335926  ],
        [-0.3018478 ],
        [-0.32836604],
        [-0.18719749],
        [-0.22514504],
        [-0.31952965]], dtype=float32), array([-1.3215524], dtype=float32))

In [21]:
def sigmoid(x):
  import math
  return 1/(1+math.exp(-x))
sigmoid(8)  



0.9996646498695336

In [22]:
X_test

T_Max     T_Min      Wind     Could  Humidity  Pressure
364   0.88  0.008850  0.047619  0.170213  0.607143  0.995619
225   0.84  0.022124  0.761905  0.563830  0.714286  0.994524
33    0.88  0.008850  0.190476  0.053191  0.071429  0.997809
100   0.88  0.013274  0.142857  0.372340  0.750000  0.992333
45    0.92  0.004425  0.142857  0.244681  0.321429  0.995619
..     ...       ...       ...       ...       ...       ...
149   0.80  0.017699  0.761905  0.808511  0.857143  0.994524
192   0.88  0.022124  0.619048  0.106383  0.500000  0.995619
287   0.88  0.013274  0.190476  0.404255  0.607143  0.996714
168   0.76  0.017699  0.809524  0.744681  0.785714  0.994524
341   0.84  0.013274  0.047619  0.585106  0.857143  0.995619

[73 rows x 6 columns]

In [23]:
def prediction_function(T_max,T_min,Wind,Could,Pressure,Humidity):
  weighted_sum = coef[0] * T_max + coef[1] * T_min + coef[2] * Wind + coef[3] * Could + coef[4] * Pressure + coef[5] * Humidity + intercept
  return sigmoid(weighted_sum)



In [24]:
prediction_function(0.88, 0.008850 , 0.047619 , 0.170213 ,0.607143,	0.995619)  

0.1069612327576373

When we input the variable values of the corresponding to the 364 raw, we can get value 0.40268136412389227. 



model.predict(X_test_scaled) line gives the same value relative to 364 raw array in 0.40268135.